src/libraries/System.Text.Json/tests/Serialization/ExtensionDataTests.cs

```diff
@@ -250,6 +250,44 @@ public static void IgnoredDataShouldNotBeExtensionData()
             Assert.Null(obj.MyOverflow);
         }
 
+
+        private class ClassWithExtensionData
+        {
+            [JsonExtensionData]
+            public Dictionary<string, object> Overflow { get; set; }
+        }
+
+        public class DictionaryOverflowConverter : JsonConverter<Dictionary<string, object>>
+        {
+            public override Dictionary<string, object> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                throw new NotImplementedException();
```

Comment:
Let's add a test for deserialization that ensures that the converter is not called.

Reply:
Prior to this PR the cache gets populated based on the converter that is specified by `[JsonConverter]` or through `JsonSerializerOptions`, but upon serialization/deserialization it is assuming that we have a `JsonDictionaryConverter<>`. My PR makes sure for writing we do not make that assumption, but for reading it is more complicated, because we store only one converter per property. If we want different behaviors for read and write, we need to have the custom serializer registered for write and `JsonDictionaryConverter<>` for read, at lease for the extension data. 

Current code
```C#
JsonConverter<object> converter = (JsonConverter<object>)state.Current.JsonPropertyInfo!.RuntimeClassInfo.ElementClassInfo!.PropertyInfoForClassInfo.ConverterBase;
```
Assumes that `ElementClassInfo` is not `null` which is not correct in this case and currently it throws with `NullReferenceException` so the suggested test would fail. I think better to fix that in a separate PR.
https://github.com/dotnet/runtime/blob/c6532913e65e168ff8a066a0cfe0ab9e5176f917/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs#L219

Reply:
We can mitigate the null ref by fetching the converter from the `Options` instance, rather than `ElementClassInfo`, similar to https://github.com/dotnet/runtime/blob/c6532913e65e168ff8a066a0cfe0ab9e5176f917/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs#L265

We should make the change in this PR.

---

The other option is to assign `ElementType` when handling `ClassType.Value` types during warm up, similar to https://github.com/dotnet/runtime/blob/c6532913e65e168ff8a066a0cfe0ab9e5176f917/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs#L170.

We probably shouldn't go down this route since `ElementType` and `ElementClassInfo` are logically reserved for collections (de)serialized without custom converters.

Reply:
In light of this discussion, let's add and test two more classes for deserialization (where the extension data element type is `JsonElement`), to ensure that the converter (another class to add)  not called.

- a test for deserialization where the converter is added to the `options.Converters` collection
- a test for deserialization where `JsonConverterAttribute` is placed on the property.

Reply:
Updated and tests are added